src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +640,337 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
 }
 
 
+/*
+ * RecursivelyPlanRecurringTupleOuterJoinWalker descends into a join tree and
+ * recursively plans all non-recurring (i.e., distributed) rels that that
+ * participate in an outer join expression together with a recurring rel,
+ * such as <distributed> in "<recurring> LEFT JOIN <distributed>", i.e.,
+ * where the recurring rel causes returning recurring tuples from the worker
+ * nodes.
+ *
+ * Returns true if given node is recurring.
+ *
+ * See RecursivelyPlanDistributedJoinNode() function for the explanation on
+ * what does it mean for a node to be "recurring" or "distributed".
+ */
+static bool
+RecursivelyPlanRecurringTupleOuterJoinWalker(Node *node, Query *query,
+											 RecursivePlanningContext *
+											 recursivePlanningContext)
+{
```

Comment:
very minor: can we order the if statements as in the same order they'd show up in the queryTree? That could make it slightly easier to follow:

if (null)
else if (FromList)
else if (JoinExpr)
else if (RangeTblRef)

or 

if (null)
else if (RangeTblRef)
else if (JoinExpr)
else if (FromList)